.github/workflows/build-and-push-docker-image.yml

```diff
@@ -0,0 +1,39 @@
+---
+
+name: 🏗️
+
+on:   # yamllint disable-line rule:truthy
+  pull_request:
+  push:
+    branches: ["release/*", "unstable/*"]
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: Docker image tag
```

Comment:
Can we incorporate the prior art from the discussions in #45 and accept the 3-segment version as an input, then deduce everything else from that + push the Git tag and advance the proper branches? Perhaps, this could also happen before this PR so it exists as a separate base.

Reply:
> Can we incorporate the prior art from the discussions in #45 and accept the 3-segment version as an input, then deduce everything else from that + push the Git tag and advance the proper branches? Perhaps, this could also happen before this PR so it exists as a separate base.

I don't understand what you're asking here. You want separate inputs for major, minor, and patch versions? What if you're building the Docker image from the `release/v1` branch? What is the "3-segment version" in that case?

Reply:
@br3ndonland I was talking about the workflow inputs. Everything else should be computed from them.

The current workflow for me is as follows:
1) Tag `v1.10.1` on top of `unstable/v1`
2) Fast-forward `release/v1` to `v1.10.1`
3) Fast-forward `release/v1.10` to `v1.10.1` (or create it)
4) `git push --atomic` everything in one go
5) Create a GitHub Release from the UI

With this, my 3-segment “input” is `1.10.1` and `v1`+`v1.10` are being extracted from that. Do you think we could bake containers into this flow with not many changes? This will involve tagging the same image with multiple tags, as I understand. And the end-users referencing versions by both tags and branches should be able to retrieve whatever container there is, right?

Reply:
@webknjaz thanks for explaining that. I think we're actually pretty close to the desired workflow. There's just a small update needed to the Docker build workflow so it pushes tags with the major and minor version numbers. I've pushed that change.

Your updated workflow would look like this:

1. Tag `v1.10.1` on top of `unstable/v1`
2. Fast-forward `release/v1` to `v1.10.1`
3. Fast-forward `release/v1.10` to `v1.10.1` (or create it)
4. `git push --atomic` everything in one go
5. Trigger the :building_construction: workflow from the Git tag. The workflow will push the Docker image with three Docker tags:
   1. `ghcr.io/pypa/gh-action-pypi-publish:v1.10.1`
   2. `ghcr.io/pypa/gh-action-pypi-publish:v1.10`
   3. `ghcr.io/pypa/gh-action-pypi-publish:v1`
6. Create a GitHub Release from the UI

Reply:
@br3ndonland so with #45, I wanted to move all these steps inside the automation. For publishing Python projects, I usually use a workflow with `workflow_dispatch` that allows me to type in the desired version to release and that workflow creates the git tag, the gh release, signs stuff and so on.

The idea is that tag+branches+gh-release represent the result of release automation being successful and aren't triggers or manual actions.

Reply:
> @br3ndonland so with #45, I wanted to move all these steps inside the automation. For publishing Python projects, I usually use a workflow with `workflow_dispatch` that allows me to type in the desired version to release and that workflow creates the git tag, the gh release, signs stuff and so on.
> 
> The idea is that tag+branches+gh-release represent the result of release automation being successful and aren't triggers or manual actions.

So... you want the Docker build workflow to run automatically when a tag is created? Or when a GitHub Release is created?

Please be more specific here, and please limit the scope of your requests. The goal of this PR is "Build Docker image and push to GHCR." The goal is not to develop your entire release workflow.

PR #45 is currently a draft and I don't think I should have to make this PR dependent on some other draft PR.

Reply:
> So... you want the Docker build workflow to run automatically when a tag is created? Or when a GitHub Release is created?

No to both. In my release processes, tags + GH Releases are the results of publishing, not the triggers. So when I trigger `workflow_dispatch` (which is a button click on GH UI + the desired version entered into the input field), the workflows create what I want to publish, send that to the primary platform (PyPI in case of Python projects or a container registry in the context of this specific PR), and only after that the same workflow creates the Git tag and the GH Release.

> Please be more specific here, and please limit the scope of your requests. The goal of this PR is "Build Docker image and push to GHCR." The goal is not to develop your entire release workflow.
> 
> PR #45 is currently a draft and I don't think I should have to make this PR dependent on some other draft PR.

I only pointed to that PR as this one seems to be touching the same topic partially. I was thinking that since you're already making a workflow that publishes the container, that could piggyback on top of this effort. I wasn't thinking of it as a dependency but rather as a replacement, but wanted to point to it for the pre-existing context around what I wanted to implement for quite a while.

This PR is attempting to change my release workflow already, after all. So I wanted to make sure it doesn't go in the direction that would complicate things, given that I already envision the release process to develop in a certain way.

If it were possible to merge the PR and keep the workflow where the Git tag is pushed, and it works right away without a container published for a while, it'd be where we could scope it. However, AFAIU, this will no longer work after merging since when there's a tag and the end-users start using it, the action will attempt pulling the image that is not yet there. There's a race condition in this process, meaning that it is prone to human error and needs to be handled carefully. And that's why I was thinking that it'd be good to integrate all the other things in a way that would keep the process robust.

P.S. Would it be useful to still give the end-users some way to force building the container instead of pulling the cached one?